config/initializers/services.rb

```diff
@@ -0,0 +1,78 @@
+# frozen_string_literal: true
+
+def assign_db(lhs, rhs)
+  if rhs.is_a? String
+    lhs.url = rhs
+  elsif rhs.respond_to?(:has_key?)
+    if rhs["url"]
+      lhs.url = rhs["url"]
+    else
+      lhs.opts = rhs
+    end
+  end
+end
+
+
+assign_db(Keycard::DB.config, Otis.config.keycard.database)
+assign_db(Checkpoint::DB.config, Otis.config.checkpoint.database)
+
+Keycard::DB.config.readonly = true if Otis.config.keycard&.readonly
+Keycard.config.access = Otis.config.keycard&.access || :direct
+
+unless Checkpoint::DB.connected?
+  if Checkpoint::DB.conn_opts.empty?
+    Checkpoint::DB.connect!(db: Sequel.sqlite)
+    Checkpoint::DB.migrate!
+  end
+end
+Checkpoint::DB.initialize!
+
+Services = Canister.new
+
+role_map = {admin: [:index, :show, :save, :edit, :create, :update],
+            staff: [:index, :show],
+            institution: [:index, :show]}
+
+Services.register(:checkpoint) do
+  Checkpoint::Authority.new(#agent_resolver: ActorAgentResolver.new,
+                            credential_resolver: Checkpoint::Credential::RoleMapResolver.new(role_map),
+                            resource_resolver: Checkpoint::Resource::Resolver.new)
+end
+
+# Initialize the Checkpoint database with admin and staff users, for compatibility with
+# the existing list in the production config.
+#
+# Checkpoint grant! allows identical database entries, hence the call
+# to permits? to try to avoid filling the DB with duplicate entries.
+# However, Checkpoint permits? returns false with identical parameters
+# to the subsequent call to grant!
+Checkpoint::DB.db[:grants].delete
+admin_role = Checkpoint::Credential::Role.new(:admin)
+staff_role = Checkpoint::Credential::Role.new(:staff)
+institution_role = Checkpoint::Credential::Role.new(:institution)
+res_wildcard = Checkpoint::Resource::AllOfAnyType.new
+res_inst = Checkpoint::Resource.new(OpenStruct.new(id: :ht_institutions))
```

Comment:
This looks like the key disconnect. I think you want an AllOfType here and in the check. As it stands, it's sort of like `OpenStruct:ht_institutions` rather than `ht_institutions:(all)`... noting that I've been using the shorthand token notation throughout. This is just &lt;type&gt;:&lt;id&gt; for each of the primitives.

Reply:
Based on this and other advice I am getting rid of the OpenStruct hack and, I think, using the correct wildcarding.

Reply:
And... yay, Travis likes this code. Ya have to poke it a couple three times.